keps/sig-node/0028-20180906-quotas-for-ephemeral-storage.md

```diff
@@ -0,0 +1,813 @@
+  from overshooting its limit; at best it catches it after the fact.
+  On a fast storage medium, such as NVMe, a pod may write 50 GB or
+  more of data before the housekeeping performed once per minute
+  catches up to it.  If the primary volume is the root partition, this
+  will completely fill the partition, possibly causing serious
+  problems elsewhere on the system.  This proposal does not address
+  this issue; _a future enforcing project would_.
+
+In many environments, these issues may not matter, but shared
+multi-tenant environments need these issues addressed.
+
+### Goals
+
+These goals apply only to local ephemeral storage, as described in
+<https://github.com/kubernetes/features/issues/361>.
+
+* Primary: improve performance of monitoring by using project quotas
+  in a non-enforcing way to collect information about storage
+  utilization of ephemeral volumes.
+* Primary: detect storage used by pods that is concealed by deleted
+  files being held open.
+* Primary: this will not interfere with the more common user and group
+  quotas.
+
+### Non-Goals
+
+* Application to storage other than local ephemeral storage.
+* Elimination of eviction as a means of enforcing ephemeral-storage
+  limits.  Pods that hit their ephemeral-storage limit will still be
+  evicted by the kubelet even if their storage has been capped by
+  enforcing quotas.
+* Enforcing node allocatable (limit over the sum of all pod's disk
+  usage, including e. g. images).
+* Enforcing limits on total pod storage consumption by any means, such
+  that the pod would be hard restricted to the desired storage limit.
+  
+### Future Work
+
+* _Enforce limits on per-volume storage consumption by using
+  enforced project quotas._
+
+## Proposal
+
+This proposal applies project quotas to emptydir volumes on qualifying
+filesystems (ext4fs and xfs with project quotas enabled).  Project
+quotas are applied by selecting an unused project ID (a 32-bit
+unsigned integer), setting a limit on space and/or inode consumption,
+and attaching the ID to one or more files.  By default (and as
+utilized herein), if a project ID is attached to a directory, it is
+inherited by any files created under that directory.
+
+_If we elect to use the quota as enforcing, we impose a quota
+consistent with the desired limit._  If we elect to use it as
+non-enforcing, we impose a large quota that in practice cannot be
+exceeded (2^63-1 bytes for XFS, 2^58-1 bytes for ext4fs).
+
+### Control over Use of Quotas
+
+At present, three feature gates control operation of quotas:
+
+* `LocalStorageCapacityIsolation` must be enabled for any use of
+  quotas.
+  
+* `FSQuotaForLSCIMonitoring` must be enabled in addition.  If this is
+  enabled, quotas are used for monitoring, but not enforcement.  At
+  present, this defaults to False, but the intention is that this will
+  default to True by initial release.
+  
+* `FSQuotaForLSCIEnforcement` must be enabled, in addition to
```

Comment:
I would prefer if we delayed adding the `FSQuotaForLSCIEnforcement` feature gate until after we have settled on a proposal for how enforcement will work.  Otherwise, what code would we place behind it?

Reply:
The low level code is very similar; see the PR.  It's not a problem to remove it if it would be problematic, although having it in there makes it possible to experiment with enforcing quotas without additional code changes.

Reply:
I've removed this from the code and moved it to future in the KEP.